lib/handler/cache-handler.js

```diff
@@ -0,0 +1,202 @@
+'use strict'
+
+const util = require('../core/util.js')
+const DecoratorHandler = require('../handler/decorator-handler')
+const {
+  parseCacheControlHeader,
+  shouldRequestBeCached,
+  parseVaryHeader
+} = require('../util/cache.js')
+
+/**
+ * Takes a request, copies its data to a cache store, and passes it to the next
+ *  handler.
+ */
+class CacheHandler extends DecoratorHandler {
+  #globalOpts
+  #req
+  #handler
+  /**
+   * @type {import('../../types/cache-interceptor.d.ts').default.CacheStoreValue | undefined}
+   */
+  #value = undefined
+
+  /**
+   * @param {import('../../types/cache-interceptor.d.ts').default.CacheOptions} globalOpts
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @param {import('../../types/dispatcher.d.ts').default.DispatchHandlers} handler
+   */
+  constructor (globalOpts, req, handler) {
+    super(handler)
+
+    this.#globalOpts = globalOpts
+    this.#req = req
+    this.#handler = handler
+  }
+
+  onHeaders (
+    statusCode,
+    rawHeaders,
+    resume,
+    statusMessage,
+    headers = util.parseHeaders(rawHeaders)
+  ) {
+    const cacheControlHeader = headers['cache-control']
+
+    // Cache control header is missing or the status code isn't what we want,
+    //  let's not try to cache this response
+    if (!cacheControlHeader || !(statusCode in this.#globalOpts.statusCodes)) {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const cacheControlDirectives = parseCacheControlHeader(cacheControlHeader)
+
+    const contentLength = headers['content-length']
+      ? Number(headers['content-length'])
+      : Infinity
+
+    const maxEntrySize = this.#globalOpts.store.maxEntrySize ?? Infinity
+
+    if (
+      maxEntrySize > contentLength &&
+      shouldRequestBeCached(cacheControlDirectives, headers.vary)
+    ) {
+      const varyDirectives = headers.vary
+        ? parseVaryHeader(headers.vary)
+        : undefined
+
+      const ttl = determineTtl(headers, cacheControlDirectives)
+      if (ttl > 0) {
+        const strippedHeaders = stripNecessaryHeaders(rawHeaders, headers)
+
+        const now = Date.now()
+        this.#value = {
+          complete: false,
+          data: {
+            statusCode,
+            statusMessage,
+            rawHeaders: strippedHeaders,
+            rawTrailers: null,
+            body: []
+          },
+          cachingDirectives: cacheControlDirectives,
+          vary: varyDirectives,
+          size: (rawHeaders?.reduce((xs, x) => xs + x.length, 0) ?? 0) +
+              (statusMessage?.length ?? 0) +
+              64,
+          cachedAt: now,
+          expiresAt: now + ttl
```

Comment:
If you already now when it should expire, it should be ok to not store the moment it was cached. i.e. use either of `cachedAt` or `expiresAt`, keeping them both seems redundant

Reply:
`cachedAt` is used for calculating the Age header